lib/swaps/Swaps.ts

```diff
@@ -0,0 +1,336 @@
+import { SwapDealRole } from '../types/enums';
+import Peer from '../p2p/Peer';
+import uuidv1 from 'uuid/v1';
+import * as packets from '../p2p/packets/types';
+import { randomBytes, createHash } from 'crypto';
+import Logger from '../Logger';
+import * as lndrpc from '../proto/lndrpc_pb';
+import LndClient from '../lndclient/LndClient';
+import Pool from '../p2p/Pool';
+import { EventEmitter } from 'events';
+import { StampedPeerOrder, StampedOwnOrder } from '../types/orders';
+
+type SwapDeal = {
+  /** The role of the local node in the swap. */
+  myRole: SwapDealRole;
+  /** The xud node pub key of the counterparty to this swap deal. */
+  peerPubKey: string;
+  /** Global order id in the XU network. */
+  orderId?: string;
+  takerDealId: string;
+  /** The number of satoshis (or equivalent) the taker is expecting to receive. */
+  takerAmount: number;
+  /** The currency the taker is expecting to receive. */
+  takerCurrency: string;
+  /** The lnd pub key of the taker. */
+  takerPubKey: string;
+  makerDealId?: string;
+  /** The number of satoshis (or equivalent) the maker is expecting to receive. */
+  makerAmount: number;
+  /** The currency the maker is expecting to receive. */
+  makerCurrency: string;
+  /** The lnd pub key of the maker. */
+  makerPubKey?: string;
+  /** The hash of the preimage. */
+  r_hash?: string;
+  r_preimage?: string;
+  createTime: number;
+  executeTime?: number;
+  competionTime?: number
+};
+
+export class Swaps extends EventEmitter {
+  private deals: SwapDeal[] = [];
+
+  constructor(private logger: Logger, private pool: Pool, private lndBtcClient: LndClient, private lndLtcClient: LndClient) {
+    super();
+
+    this.bind();
+  }
+
+  private bind() {
+    this.pool.on('packet.dealRequest', this.handleDealRequest);
+    this.pool.on('packet.dealResponse', this.handleDealResponse);
+    this.pool.on('packet.swapRequest', this.handleSwapRequest);
+    this.pool.on('packet.swapResponse', this.handleSwapResponse);
+  }
+
+  private getDeal = (role: SwapDealRole, dealId: string): SwapDeal | undefined => {
+    for (const deal of this.deals) {
+      if (role === SwapDealRole.Maker && deal.makerDealId === dealId) {
+        return deal;
+      }
+      if (role === SwapDealRole.Taker && deal.takerDealId === dealId) {
+        return deal;
+      }
+    }
+    return undefined;
+  }
+
+  public addDeal = (deal: SwapDeal) => {
+    this.deals.push(deal);
+  }
+
+  public getDealByHash = (r_hash: string): SwapDeal | undefined => {
```

Comment:
What about using a map instead of an array to make this method way more efficient?

Reply:
I agree, I just didn't get to it in this PR. I'm thinking going forward there should be a single global deal id that both maker and taker use the way we do order ids (instead of each having their own) and then a mapping of dealId to deal and r_hash to deal.